### PR TITLE
[BEAM-7389] Add code examples for Partition page

### PR DESCRIPTION
**Pull request by davidcavazos**
Adding code samples for the `Partition` page.

R: @aaltay @rosetn Can you take a look at this whenever you have a chance? Thanks!

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by davidcavazos** [comment]
Staged: http://apache-beam-website-pull-requests.storage.googleapis.com/9261/documentation/transforms/python/elementwise/partition/index.html

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 1]
```diff
@@ -39,12 +46,110 @@ You cannot determine the number of partitions in mid-pipeline
...
+<br>
+
+### Example 3: Partition with multiple arguments
+
+You can pass functions with multiple arguments to `Partition`.
+They are passed as additional positional arguments or keyword arguments to the function.
+
+In this example, `split_dataset` takes `plant`, `num_partitions`, and `ratio` as arguments.
+`num_partitions` is used by `Partitions` as a positional argument,
+while any other argument will be passed to `split_dataset`.
```
Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: Can you add some more explanation by using the concrete details? For example, what does sample will do with the arguments in beam.Partition(split_dataset, 2, ratio=[8, 2]) and how that affects the output. 

"test_dataset" and "train_dataset" might make sense to some, but isn't generic enough. Either rename them to something like "dataset1" and "dataset2" or in your explanation before the code.

Something along the lines of "In this example, we want to split the dataset into a training dataset and a testing dataset using `Partition`. " And elaborate on what setting the ration to [8, 2] does.

Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Got it, I just added a more detailed explanation.

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 2]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+
+In machine learning, it is common to split it into
+[training and a testing datasets](https://en.wikipedia.org/wiki/Training,_validation,_and_test_sets).
+Typically, 80% of the data is used for training a model, and 20% is used for testing.
+
+We will split a `PCollection` dataset into training and testing datasets.
+We define `split_dataset` which receives the element, the number of partitions, and an additional argument `ratio` that describes the ratio of the split.
```
Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: add comma 

"`split_dataset`, which"

Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: "and an additional argument `ratio` that describes the ratio of the split."->"and an additional argument `ratio`, which is a list of numbers that represents that ratio of items in the partitions."

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 3]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+The `ratio` is a list of numbers which represents the ratio how many items will go into each partition.
```
Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: Delete this sentence if you choose the rewrite above

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 4]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+If we want an 80%/20% split, we can specify a ratio of `[8, 2]` which means that for every 10 elements, 8 will go into the first partition and 2 will go into the second.
```
Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: Add comma  "`[8, 2]`, which"

"8 will go into the first partition and 2 will go into the second."->"8 go into the first partition and 2 go into the second."

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 5]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+
+To decide which partition to send each element, we'll have different buckets.
```
Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: "In order to determine which partition to send each element, we have different buckets."

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 6]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+For our case `[8, 2]` will have *8+2*=**10** buckets, where the first 8 buckets represent the first partition and the last 2 buckets represent the second partition.
```
Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: "For our case `[8, 2]` will have *8+2*=**10** buckets,"->"For our case, `[8, 2]`  has **10** buckets,"

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 7]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+
+First, we need to make sure that the ratio's length corresponds to the `num_partitions` we pass.
```
Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: "need to make sure"-"check"

Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: "ratio's length"->"ratio list's length"

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 8]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+We then get a bucket index for each element, in the range from 0 to 10 (`num_buckets-1`).
+We could do `hash(element) % len(ratio)`, but instead we'll sum all the ASCII characters of the JSON representation to make it deterministic.
```
Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: "we’ll"->"we"

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 9]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+Finally, we loop through all the elments in the ratio and have a running total to identify the partition index to which that bucket corresponds.
```
Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: "elments"->"elements"

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 10]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+<br>
+
+### Example 3: Partition with multiple arguments
+
+You can pass functions with multiple arguments to `Partition`.
+They are passed as additional positional arguments or keyword arguments to the function.
+
+In this example, `split_dataset` takes `plant`, `num_partitions`, and `ratio` as arguments.
+`num_partitions` is used by `Partitions` as a positional argument,
+while any other argument will be passed to `split_dataset`.
```
[the remark this answers is not in the document]
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Sounds good, changed

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 11]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+
+In machine learning, it is common to split it into
```
[the remark this answers is not in the document]
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Thanks, changed

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 12]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+[training and a testing datasets](https://en.wikipedia.org/wiki/Training,_validation,_and_test_sets).
+Typically, 80% of the data is used for training a model, and 20% is used for testing.
```
[the remark this answers is not in the document]
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Removed

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 13]
```diff
@@ -39,12 +46,130 @@ You cannot determine the number of partitions in mid-pipeline
...
+
+We will split a `PCollection` dataset into training and testing datasets.
```
[the remark this answers is not in the document]
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: It was a bit repetitive with the paragraph above that also started like that. That same paragraph had almost the same information but more detailed, so I ended up moving it here and replacing this line.

File: website/src/documentation/transforms/python/element-wise/partition.md (hunk 2 above)
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Done

Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Ended up doing something slightly different

File: website/src/documentation/transforms/python/element-wise/partition.md (hunk 3 above)
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Decided to keep it as a separate sentence to make them shorter and easier to read.

File: website/src/documentation/transforms/python/element-wise/partition.md (hunk 4 above)
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Done

File: website/src/documentation/transforms/python/element-wise/partition.md (hunk 6 above)
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Done

File: website/src/documentation/transforms/python/element-wise/partition.md (hunk 7 above)
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Done

Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Done

File: website/src/documentation/transforms/python/element-wise/partition.md (hunk 8 above)
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Done

File: website/src/documentation/transforms/python/element-wise/partition.md (hunk 9 above)
Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Done

File: website/src/documentation/transforms/python/element-wise/partition.md [hunk 14]
```diff
@@ -39,12 +46,133 @@ You cannot determine the number of partitions in mid-pipeline
...
+They are passed as additional positional arguments or keyword arguments to the function.
+
+In machine learning, it is a common task to split data into
+[training and a testing datasets](https://en.wikipedia.org/wiki/Training,_validation,_and_test_sets).
+Typically, 80% of the data is used for training a model and 20% is used for testing.
+
+In this example, we split a `PCollection` dataset into training and testing datasets.
+We define `split_dataset`, which takes the `plant` element, `num_partitions`,
+and an additional argument `ratio`.
+The `ratio` is a list of numbers which represents the ratio how many items will go into each partition.
```
Comment (rosetn) [inline website/src/documentation/transforms/python/element-wise/partition.md]: "the ratio how"->"the ratio of how "

Comment (davidcavazos) [reply website/src/documentation/transforms/python/element-wise/partition.md]: Done!

**Comment by aaltay** [comment]
LGTM. Please ping me once you get approval from @rosetn .

**Pull request merged**